core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java

```diff
@@ -352,64 +378,85 @@ private void runInferencingSparqlQueries(SailRepositoryConnection shaclSailConne
 	 *
 	 * @throws SailException if the thread is interrupted while waiting to obtain the lock.
 	 */
-	long acquireExclusiveWriteLock(long stamp) {
-		if (lock.validate(stamp)) {
-			return stamp;
+	Lock acquireExclusiveWriteLock(Lock lock) {
+
+		if (lock != null && lock.isActive()) {
+			return lock;
 		}
 
+		assert lock == null;
+
 		if (threadHoldingWriteLock == Thread.currentThread()) {
 			throw new SailConflictException(
 					"Deadlock detected when a single thread uses multiple connections " +
 							"interleaved and one connection has modified the shapes without calling commit() " +
 							"while another connection also tries to modify the shapes!");
 		}
 
-		long newStamp = lock.writeLock();
-		threadHoldingWriteLock = Thread.currentThread();
-		return newStamp;
-	}
-
-	boolean holdsWriteLock(long stamp) {
-		return lock.validate(stamp);
+		try {
+			Lock writeLock = lockManager.getWriteLock();
+			threadHoldingWriteLock = Thread.currentThread();
+			return writeLock;
+		} catch (InterruptedException e) {
+			throw new RuntimeException(e);
```

Comment:
I don't really know what I should do with the interrupt exception. This should only happen if the users code calls interrupt, so I reckon the safest thing is just to re throw it. @jeenbroekstra do you have any recommendations?